components/locid_transform/src/directionality.rs

```diff
@@ -0,0 +1,150 @@
+/// assert_eq!(ld.get(&locale), Some(Direction::LeftToRight));
+/// ```
+///
+/// [`CLDR`]: http://cldr.unicode.org/
+#[derive(Debug)]
+pub struct LocaleDirectionality {
+    script_direction: DataPayload<ScriptDirectionV1Marker>,
+    expander: LocaleExpander,
+}
+
+impl LocaleDirectionality {
+    /// A constructor which takes a [`DataProvider`] and creates a [`LocaleDirectionality`].
+    ///
+    /// [📚 Help choosing a constructor](icu_provider::constructors)
+    /// <div class="stab unstable">
+    /// ⚠️ The bounds on this function may change over time, including in SemVer minor releases.
+    /// </div>
+    pub fn try_new_unstable<P>(provider: &P) -> Result<LocaleDirectionality, LocaleTransformError>
+    where
+        P: DataProvider<ScriptDirectionV1Marker>
+            + DataProvider<LikelySubtagsForLanguageV1Marker>
+            + DataProvider<LikelySubtagsForScriptRegionV1Marker>
+            + ?Sized,
+    {
+        let expander = LocaleExpander::try_new_unstable(provider)?;
+        Self::try_new_with_expander_unstable(provider, expander)
+    }
+
+    icu_provider::gen_any_buffer_constructors!(
+        locale: skip,
+        options: skip,
+        error: LocaleTransformError
+    );
+
+    /// Creates a [`LocaleDirectionality`] with a custom [`LocaleExpander`] object.
+    pub fn try_new_with_expander_unstable<P>(
+        provider: &P,
+        expander: LocaleExpander,
+    ) -> Result<LocaleDirectionality, LocaleTransformError>
+    where
+        P: DataProvider<ScriptDirectionV1Marker> + ?Sized,
+    {
+        let script_direction: DataPayload<ScriptDirectionV1Marker> =
+            provider.load(Default::default())?.take_payload()?;
+
+        Ok(LocaleDirectionality {
+            script_direction,
+            expander,
+        })
+    }
+
+    /// Returns the script direction of the given locale.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_locid::locale;
+    /// use icu_locid_transform::{Direction, LocaleDirectionality};
+    ///
+    /// let ld = LocaleDirectionality::try_new_unstable(&icu_testdata::unstable())
+    ///     .expect("create failed");
+    ///
+    /// let locale = locale!("en-US");
+    /// assert_eq!(ld.get(&locale), Some(Direction::LeftToRight));
+    ///
+    /// let locale = locale!("ar");
+    /// assert_eq!(ld.get(&locale), Some(Direction::RightToLeft));
+    ///
+    /// let locale = locale!("fr-Brai-FR");
+    /// assert_eq!(ld.get(&locale), None);
+    /// ```
+    pub fn get(&self, locale: &Locale) -> Option<Direction> {
+        let script = locale.id.script.or_else(|| {
+            let expander = self.expander.as_borrowed();
+            match (locale.id.language.is_empty(), locale.id.region) {
+                (false, Some(region)) => expander.get_lr(locale.id.language, region),
+                (false, None) => expander.get_l(locale.id.language).map(|(s, _)| s),
+                (true, Some(region)) => expander.get_r(region).map(|(_, s)| s),
```

Comment:
You can use `Language::UND` as a pattern:

```suggestion
            match (locale.id.language, locale.id.region) {
                (Language::UND, Some(region)) => expander.get_r(region).map(|(_, s)| s),
                (lang, Some(region)) => expander.get_lr(lang, region),
                (lang, None) => expander.get_l(lang).map(|(s, _)| s),
```

Reply:
That looks better, thanks! Your version potentially calls `get_l` with `Language::UND`, I assume that's alright?

Reply:
You can move the `_` case up as `(Language::UND, None)` to avoid the lookup.